OpenRA.Game/Exts.cs

```diff
@@ -113,19 +113,25 @@ public static int IndexOf<T>(this T[] array, T value)
 
 		public static T Random<T>(this IEnumerable<T> ts, MersenneTwister r)
 		{
-			var xs = ts as ICollection<T>;
-			if (xs != null)
-				return xs.ElementAt(r.Next(xs.Count));
-			var ys = ts.ToList();
-			return ys[r.Next(ys.Count)];
+			return Random(ts, r, true);
 		}
 
 		public static T RandomOrDefault<T>(this IEnumerable<T> ts, MersenneTwister r)
 		{
-			if (!ts.Any())
-				return default(T);
+			return Random(ts, r, false);
+		}
 
-			return ts.Random(r);
+		static T Random<T>(IEnumerable<T> ts, MersenneTwister r, bool throws)
+		{
+			var xs = ts as ICollection<T>;
+			xs = xs ?? ts.ToList();
```

Comment:
Actually the problem is that we need to know the size of the collection. If we don't do that we need to do a LINQ ```Count()``` to get the size and then ```ElementAt()```. That enumerates the input twice which is inefficient and leads to the bug we are trying to fix.

If we have a ```ICollection```, then we can use the ```Count``` property which solves our problem of getting the size without having to enumerate a second time. If we don't have a ```ICollection```, then we use ```ToList``` to get us one.